compiler/noirc_evaluator/src/ssa/opt/independent_subgraph_detection.rs

```diff
@@ -0,0 +1,383 @@
+//! This module defines an SSA pass that detects if the final function has any subgraphs independent from inputs and outputs.
+//! If this is the case, then part of the final circuit can be completely replaced by any other passing circuit, since there are no constraints ensuring connections.
+//! So the compiler informs the developer of this as a bug
+use im::HashMap;
+
+use crate::errors::{InternalBug, SsaReport};
+use crate::ssa::ir::basic_block::BasicBlockId;
+use crate::ssa::ir::function::RuntimeType;
+use crate::ssa::ir::function::{Function, FunctionId};
+use crate::ssa::ir::instruction::{Instruction, InstructionId, Intrinsic};
+use crate::ssa::ir::value::{Value, ValueId};
+use crate::ssa::ssa_gen::Ssa;
+use std::collections::{BTreeMap, HashSet};
+
+impl Ssa {
+    /// Go through each top-level non-brillig function and detect if it has independent subgraphs
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn detect_independent_subgraphs(mut self) -> Ssa {
+        for function in self.functions.values() {
+            match function.runtime() {
+                RuntimeType::Acir { .. } => {
+                    let warnings = self.detect_independent_subgraphs_within_function(function);
+                    self.warnings.extend(warnings);
+                }
+                RuntimeType::Brillig => (),
+            }
+        }
+        self
+    }
+
+    /// Detect independent subgraphs (not connected to function inputs or outputs) and return a vector of bug reports if some are found
+    fn detect_independent_subgraphs_within_function(&self, function: &Function) -> Vec<SsaReport> {
+        let mut context = Context::default();
+
+        // Go through each block in the function and create a list of sets of ValueIds connected by instructions
+        context.block_queue.push(function.entry_block());
+        while let Some(block) = context.block_queue.pop() {
+            if context.visited_blocks.contains(&block) {
+                continue;
+            }
+            context.visited_blocks.insert(block);
+            context.connect_value_ids_in_block(function, block, &self.functions);
+        }
+
+        // Merge ValueIds into sets, where each original small set of ValueIds is merged with another set if they intersect
+        context.merge_sets();
+
+        let function_parameters = function.parameters();
+        let variable_parameters_and_return_values: HashSet<ValueId> = function_parameters
+            .iter()
+            .chain(function.returns())
+            .filter(|&x| match function.dfg[*x] {
+                Value::NumericConstant { .. } => false, // Constant values don't connect elements and can be reused in different subgraphs, so we need to avoid them
+                _ => true,
+            })
+            .copied()
+            .collect();
```

Comment:
```suggestion
            .copied();
```
Doesn't look like we need to `collect` here either. (Also need to remove the `: HashSet<ValueId>` annotation above)

Reply:
Done